components/Dialogs/tests/snapshot/MDCAlertControllerDynamicTypeTests.m

```diff
@@ -0,0 +1,280 @@
+// Copyright 2019-present the Material Components for iOS authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#import "MaterialSnapshot.h"
+
+#import <CoreGraphics/CoreGraphics.h>
+#import <UIKit/UIKit.h>
+
+#import "MaterialDialogs.h"
+#import "MaterialTypography.h"
+
+/** A @c MDCAlertController test fake to override the @c traitCollection to test for dynamic type.
+ */
+@interface AlertControllerDynamicTypeSnapshotTestFake : MDCAlertController
+@property(nonatomic, strong) UITraitCollection *traitCollectionOverride;
+@end
+
+@implementation AlertControllerDynamicTypeSnapshotTestFake
+
+- (UITraitCollection *)traitCollection {
+  return self.traitCollectionOverride ?: [super traitCollection];
+}
+
+@end
+
+@interface MDCAlertControllerDynamicTypeTests : MDCSnapshotTestCase
+@property(nonatomic, strong, nullable) AlertControllerDynamicTypeSnapshotTestFake *alertController;
+@end
+
+@implementation MDCAlertControllerDynamicTypeTests
+
+- (void)setUp {
+  [super setUp];
+
+  // Uncomment below to recreate all the goldens (or add the following line to the specific
+  // test you wish to recreate the golden for).
+  //  self.recordMode = YES;
+
+  self.alertController = [[AlertControllerDynamicTypeSnapshotTestFake alloc] init];
+  self.alertController.title = @"Material";
+  self.alertController.message =
+      @"Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt "
+      @"ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation "
+      @"ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in "
+      @"reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur "
+      @"sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id "
+      @"est laborum.";
+  MDCAlertAction *fakeAction = [MDCAlertAction actionWithTitle:@"Foo"
+                                                       handler:^(MDCAlertAction *action){
+                                                       }];
+  [self.alertController addAction:fakeAction];
+  self.alertController.mdc_adjustsFontForContentSizeCategory = YES;
+  self.alertController.adjustsFontForContentSizeCategoryWhenScaledFontIsUnavailable = NO;
```

Comment:
If these are DT snapshot tests, why are you turning off DT scaling?

Reply:
This is pre work because in a follow up we are going to add dynamic type scaling curves then the snapshot should update.

Reply:
Ok, then I think only the naming issue and missing tests for the current/legacy implementation of DT remains.

Reply:
Are you asking for test for when mdc_adjustFontForContentSizeCategory is true and adjustFontForContentSizeCategoryWhenScaledFontIsUnavailable if also true?

Reply:
Yep, that would match the description of the PR.  I believe those tests would change as well when the new support is added.

Reply:
There's also tests to be written for scaling WITHOUT the scalable fonts, though these are for the legacy code and I won't insist on them.

-  [ ] Scalable font + automatic scaling + legacy off
-  [ ] Scalable font + automatic scaling + legacy on
-  [ ] Unscalable font + automatic scaling + legacy off (optional)
-  [ ] Unscalable font + automatic scaling + legacy on (optional)

Reply:
Unscalable font test I'll do in a follow up to keep this as only snapshot test.

Reply:
In order to simplify the matrix I'll say there are these options since _automatic scaling_ is part of all of them

1. scalable font + legacy off
2. scalable font + legacy on
3. unscalable font + legacy off
4.  unscalable font + legacy on

The logical flow would be
```
if (mdc_adjustsFontForContentSizeCategory) {
  if (font.mdc_scalingCurve) {
      // Path 1 & 2 legacy never impacts the path if scalable fonts are on
  } else {
     if (mdc_adjustsFontForContentSizeCategoryWhenScaledFontIsUnavailable) {
      // Path 3
    }
      // Path 4
  }
}
This PR adds test for paths 1 & 2 in a follow up I'll add unit test for paths 3 & 4.

Reply:
Please at least one test with `adjustsFontForContentSizeCategoryWhenScaledFontIsUnavailable = YES` to demonstrate that the logic was implemented correctly.